android/src/main/java/ee/forgr/capacitor_updater/DownloadService.java

```diff
@@ -54,52 +44,104 @@ protected void onHandleIntent(Intent intent) {
-          final byte[] buffer = new byte[bufferSize];
-          int length;
-
-          int bytesRead = bufferSize;
-          int percent = 0;
-          this.notifyDownload(id, 10);
-          while ((length = dis.read(buffer)) > 0) {
-            fos.write(buffer, 0, length);
-            final int newPercent = (int) ((bytesRead * 100) / totalLength);
-            if (totalLength > 1 && newPercent != percent) {
-              percent = newPercent;
-              this.notifyDownload(id, this.calcTotalPercent(percent, 10, 70));
-            }
-            bytesRead += length;
+        String contentType = httpConn.getContentType();
+        int contentLength = httpConn.getContentLength() + (int) downloadedBytes;
+
+        InputStream inputStream = httpConn.getInputStream();
+        FileOutputStream outputStream = new FileOutputStream(
+          tempFile,
+          downloadedBytes > 0
+        );
+
+        // Writing initial progression into file
+        if (downloadedBytes == 0) {
+          try (
+            BufferedWriter writer = new BufferedWriter(
+              new FileWriter(progressFile)
+            )
+          ) {
+            writer.write(String.valueOf(downloadedBytes));
+          }
+        }
+
+        int bytesRead = -1;
+        byte[] buffer = new byte[4096];
+        int lastPercent = 0;
+        while ((bytesRead = inputStream.read(buffer)) != -1) {
+          outputStream.write(buffer, 0, bytesRead);
+          downloadedBytes += bytesRead;
+
+          // Updating the progress file
+          try (
+            BufferedWriter writer = new BufferedWriter(
+              new FileWriter(progressFile)
+            )
+          ) {
+            writer.write(String.valueOf(downloadedBytes));
+          }
+
+          // Saving progress (flushing every 100 Ko)
+          if (downloadedBytes % 102400 == 0) {
+            outputStream.flush();
+          }
+
+          // Computing percentage
+          int percent = calcTotalPercent(downloadedBytes, contentLength);
+          if (percent != lastPercent) {
+            notifyDownload(id, percent);
+            lastPercent = percent;
           }
-          publishResults(dest, id, version, checksum, sessionKey, "");
         }
+
+        outputStream.close();
+        inputStream.close();
+
+        // Rename the temp file with the final name (dest)
+        tempFile.renameTo(new File(documentsDir, dest));
+        progressFile.delete();
+        publishResults(dest, id, version, checksum, sessionKey, "");
+      } else {
+        progressFile.delete();
       }
-    } catch (OutOfMemoryError e) {
-      e.printStackTrace();
-      publishResults("", id, version, checksum, sessionKey, "low_mem_fail");
-    } catch (Exception e) {
+      httpConn.disconnect();
+    } catch (IOException e) {
       e.printStackTrace();
-      publishResults(
-        "",
-        id,
-        version,
-        checksum,
-        sessionKey,
-        e.getLocalizedMessage()
-      );
```

Comment:
Why did you remove this code?

Reply:
Yup mb, forgot to bring back out of memory handler while rewriting the code logic, brought it back in [33d9300](https://github.com/Cap-go/capacitor-updater/pull/400/commits/33d9300db909c8fa3bf8158c61251b89b581e396)